pkg/cmd/pipeline/list.go

```diff
@@ -16,51 +16,172 @@ package pipeline
 
 import (
 	"fmt"
+	"io"
 	"os"
+	"text/tabwriter"
 
 	"github.com/spf13/cobra"
 	"github.com/tektoncd/cli/pkg/cli"
+	"github.com/tektoncd/cli/pkg/formatted"
+	"github.com/tektoncd/cli/pkg/printer"
 	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
-	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	cliopts "k8s.io/cli-runtime/pkg/genericclioptions"
 )
 
-func ListCommand(p cli.Params) *cobra.Command {
+const (
+	emptyMsg string = "No pipelines found"
+	header   string = "NAME\tAGE\tLAST RUN\tSTARTED\tDURATION\tSTATUS"
+	body     string = "%s\t%s\t%s\t%s\t%s\t%s\n"
+)
 
-	defaultOutput := `jsonpath={range .items[*]}{.metadata.name}{"\n"}{end}`
-	f := cliopts.NewPrintFlags("").WithDefaultOutput(defaultOutput)
+func listCommand(p cli.Params) *cobra.Command {
+	f := cliopts.NewPrintFlags("list")
 
 	c := &cobra.Command{
 		Use:     "list",
 		Aliases: []string{"ls"},
 		Short:   "Lists pipelines in a namespace",
 		Long:    ``,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			ps, err := list(p)
-
+			output, err := cmd.LocalFlags().GetString("output")
 			if err != nil {
-				fmt.Fprintf(os.Stderr, "Failed to list pipelines from %s namespace\n", p.Namespace())
+				fmt.Fprint(os.Stderr, "Error: output option not set properly \n")
 				return err
 			}
 
-			printer, err := f.ToPrinter()
-			if err != nil {
-				return err
+			if output != "" {
+				return printPipelineListObj(cmd.OutOrStdout(), p, f)
 			}
-			return printer.PrintObj(ps, cmd.OutOrStdout())
+
+			return printPipelineDetails(cmd.OutOrStdout(), p)
 		},
 	}
 	f.AddFlags(c)
 
 	return c
 }
 
-func list(p cli.Params) (*v1alpha1.PipelineList, error) {
+func printPipelineDetails(out io.Writer, p cli.Params) error {
+
+	ps, prs, err := listPipelineDetails(p)
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to list pipelines from %s namespace\n", p.Namespace())
+		return err
+	}
+
+	if len(ps.Items) == 0 {
+		fmt.Fprintln(out, emptyMsg)
+		return nil
+	}
+
+	w := tabwriter.NewWriter(out, 0, 5, 3, ' ', tabwriter.TabIndent)
+	defer w.Flush()
+
+	fmt.Fprintln(w, header)
+
+	for _, pipeline := range ps.Items {
+		pr, ok := prs[pipeline.Name]
+		if !ok {
+			fmt.Fprintf(w, body,
+				pipeline.Name,
+				formatted.Age(pipeline.CreationTimestamp, p.Time()),
+				"---",
```

Comment:
Maybe we could extract this in constants as well?

Reply:
yes, I can although, `"---"` says better than `empty`